File: app/src/main/java/com/mycompany/myapp/ui/main/MainFragment.java
package com.mycompany.myapp.ui.main;

import android.content.Context;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.mycompany.myapp.CommitItemBinding;
import com.mycompany.myapp.R;
import com.mycompany.myapp.ui.BaseFragment;
import com.mycompany.myapp.ui.main.MainPresenter.CommitView;
import com.mycompany.myapp.util.recyclerview.ArrayAdapter;

import javax.inject.Inject;

public class MainFragment extends BaseFragment {
    public interface MainFragmentHost {
        void inject(MainFragment fragment);
    }

    @Inject MainPresenter presenter;
    private MainFragmentBinding binding;
    private MainFragmentHost host;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        host = (MainFragmentHost) context;
        host.inject(this);
    }

    @Override
    public void onDetach() {
        host = null;
        super.onDetach();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_main, container, false);
        binding.setPresenter(presenter);

        binding.commits.setLayoutManager(new LinearLayoutManager(getActivity()));
        binding.commits.setAdapter(new CommitsAdapter());

        return binding.getRoot();
    }

    @Override
    public void onDestroyView() {
        binding.unbind();
        super.onDestroyView();
    }

    private static class CommitsAdapter extends ArrayAdapter<CommitView, CommitViewHolder> {
        @Override
        public CommitViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());
            CommitItemBinding commitItemBinding = CommitItemBinding.inflate(layoutInflater, parent, false);
            return new CommitViewHolder(commitItemBinding);
        }

        @Override
        public void onBindViewHolder(CommitViewHolder holder, int position) {
            CommitView commit = getItemAtPosition(position);
            holder.bind(commit);
        }
    }

    static class CommitViewHolder extends RecyclerView.ViewHolder {
        private final CommitItemBinding binding;

        private CommitViewHolder(CommitItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        public void bind(CommitView item) {
            binding.setItem(item);
            binding.executePendingBindings();
        }
    }
}

File: app/src/main/java/com/mycompany/myapp/ui/main/MainFragment.kt
package com.mycompany.myapp.ui.main

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.snackbar.Snackbar
import com.mycompany.myapp.CommitItemBinding
import com.mycompany.myapp.data.api.github.model.Commit
import com.mycompany.myapp.databinding.FragmentMainBinding
import com.mycompany.myapp.ui.BaseFragment
import com.mycompany.myapp.ui.SimpleSnackbarMessage
import com.mycompany.myapp.util.recyclerview.ArrayAdapter

class MainFragment : BaseFragment() {
    private lateinit var viewModel: MainViewModel
    private lateinit var binding: FragmentMainBinding

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        viewModel = getViewModel(MainViewModel::class)

        binding = FragmentMainBinding.inflate(inflater, container, false)
        binding.vm = viewModel

        viewModel.snackbarMessage.observe(this, object : SimpleSnackbarMessage.SnackbarObserver {
            override fun onNewMessage(message: String) {
                Snackbar.make(binding.root, message, Snackbar.LENGTH_LONG).show()
            }
        })

        (activity as AppCompatActivity).let {
            it.setSupportActionBar(binding.toolbar)
            binding.toolbar.title = it.packageName
        }

        binding.commits.layoutManager = LinearLayoutManager(activity)
        binding.commits.adapter = CommitsAdapter()

        return binding.root
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        viewModel.saveState(outState)
    }

    override fun onResume() {
        super.onResume()
        viewModel.onResume()
    }

    override fun onDestroyView() {
        binding.unbind()
        super.onDestroyView()
    }

    private class CommitsAdapter : ArrayAdapter<Commit, CommitViewHolder>() {
        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CommitViewHolder {
            val layoutInflater = LayoutInflater.from(parent.context)
            val binding = CommitItemBinding.inflate(layoutInflater, parent, false)
            return CommitViewHolder(binding)
        }

        override fun onBindViewHolder(holder: CommitViewHolder, position: Int) {
            val commit = getItemAtPosition(position)
            holder.bind(commit)
        }
    }

    private class CommitViewHolder(
            private val binding: CommitItemBinding)
        : RecyclerView.ViewHolder(binding.root) {

        fun bind(item: Commit) {
            binding.item = item
            binding.executePendingBindings()
        }
    }
}
